File: Data Structures & Algorithms/duplicate-integer/submission-1.py
class Solution: 
    def hasDuplicate(self, nums: List[int]) -> bool:
        '''psuedocode logic: 
        - if element in nums not in hash set 'seen', add to hash map 
        - if element in nums in hash set 'seen', return true 
        - else return false 
        '''
        #1. BRUTE FORCE - O(n^2) time, O(1) space 
        '''for i in range (len(nums)):
            for j in range (i+1, len(nums)):
                if nums[i] == nums[j]:
                    return True 
        return False'''

        #2. SORTING - O(n log n) time, O(1) space
        '''nums.sort()
        for i in range(1, len(nums)):
            if nums[i] == nums[i-1]:
                return True 
        return False'''

        #3. HASH SET - O(n) time, O(n) space 
        seen = set()
        for element in nums: 
            if element in seen: 
                return True 
            seen.add(element)
        return False 


        

File: Data Structures & Algorithms/merge-intervals/submission-1.java
public class Solution {
    public int[][] merge(int[][] intervals) {
        /**1. restate
        - input: list of intervals (might have some that overlap each other)
        - output: merge the ones that overlap 
        - key observations: sorting the list by start time will ensure that
            if there is an overlap in intervals, it will overlap with
            the interval(s) adjacent to it 
            - we do not know the length of the output array until all intervals 
            processed; use arraylist then convert to array 
        */

        /**2. example walkthrough
        i = [1, 3] j = [1, 5] --> overlapping? yes, f_i >= s_j ? true 
            merge: [s_i, f_j] = [1, 5]
        new list: [1, 5] [6, 7] 
        i = [1, 5] j = [6 ,7] -> overlapping? no, f_i >= s_j ? false 

        return [1, 5] [6, 7]
        */

        /**3. brute force
        - add all intervals to arraylist 
        - for each interval, check all other intervals for overlapping 
        - merge accordingly 
        - remove unmerged intervals; add merged interval
        - convert final arraylist into array 
            - new int [] size  arraylist.size()
            - for int[] in arraylist -> int[i] = arraylist.get(i)
        - time O(n^2)
        */

        /**4. derive better approach
        - sort intervals by start time
        - for two adjacent intervals 
            - check if overlapping 
            - if yes? 1) merge 2) add merged to list 3) remove two unmerged from list
        */ 

        /**5. complexity 
        O(n log n)
        */ 

        // CODE

        List<int []> output = new ArrayList<>(); 

        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));

        output.add(intervals[0]); // just add the first interval

        for (int[] interval : intervals) {
            int[] last = output.get(output.size() - 1);
            if (interval[0] <= last[1]) {
                last[1] = Math.max(last[1], interval[1]);
            } 
            else {
                output.add(interval);
            }
        }

        // building the output 
        int[][] res = new int[output.size()][];
        for (int i = 0; i < output.size(); i++) {
            res[i] = output.get(i);
        }
        return res;

    }
}